include/circt/Scheduling/Problems.h

```diff
@@ -268,6 +268,82 @@ class CyclicProblem : public virtual Problem {
   virtual LogicalResult verify() override;
 };
 
+/// This class models the accumulation of physical propagation delays on
+/// combinational paths along SSA dependences within a scheduling problem's
+/// abstract time steps/cycles.
+///
+/// Each operator type is annotated with estimated values for incoming and
+/// outgoing delays. Combinational operators (zero-latency, no internal
+/// registers) have only a single propagation delay; this important special case
+/// is modeled by setting the incoming and outgoing delays to the same value.
+///
+/// A solution to this problem comprises physical start times for each operation
+/// that satisfy a given cycle time constraint.
+class ChainingProblem : public virtual Problem {
+private:
+  ProblemProperty<float> cycleTime;
+  OperatorTypeProperty<float> incomingDelay, outgoingDelay;
+  OperationProperty<float> physicalStartTime;
+
+public:
+  explicit ChainingProblem(Operation *containingOp) : Problem(containingOp) {}
+
+  /// The cycle time constrains the accumulated propagation delays along
+  /// combinational paths inside a time step.
+  Optional<float> getCycleTime() { return cycleTime; }
+  void setCycleTime(float time) { cycleTime = time; }
+
+  /// The incoming delay denotes the propagation time from the operand inputs to
+  /// either the result outputs (combinational operators) or the first internal
+  /// register stage.
+  Optional<float> getIncomingDelay(OperatorType opr) {
+    return incomingDelay.lookup(opr);
+  }
+  void setIncomingDelay(OperatorType opr, float delay) {
+    incomingDelay[opr] = delay;
+  }
```

Comment:
I think it's important to model 2 things here:  
1) operations with different widths will have different delays.  It's not clear to me how you intend to model this.   will you have different operator types?  
2) operations with multiple inputs and outputs almost always have different delays and/or latencies.  how do you plan to model this?

Reply:
1. Yes, in the current design of the infra, operator types are completely unaware of the (MLIR) operations/types of the (scheduling problem's) operations they are attached to. So the client code has to set up different operator types for the different bitwidths.
2. The problem model (or more accurately, the `Dependence` class) is prepared for that, but the currently existing `Problem`s don't use operand/result indices yet. My plan was to introduce new trait-like subproblems in which the operator types carry per-input/output latencies and delays in separate PRs.